### PR TITLE
fix for when in algorithms

### PR DESCRIPTION
**Pull request by vwaurich**

### CODE REVIEW

**Comment by lochel** [comment]
@vwaurich What issue did you fix?

File: Compiler/BackEnd/FindZeroCrossings.mo [hunk 1]
```diff
@@ -176,20 +176,27 @@ algorithm
...
-      alg_ = DAE.ALGORITHM_STMTS(stmts);
-      eqn = BackendDAE.ALGORITHM(size, alg_, source, crefExpand, attr);
-      equationArray = BackendEquation.addEquation(eqn, equationArray);
+    case (BackendDAE.ALGORITHM(size=0, alg=alg_, source=source, expand=crefExpand, attr=attr), (equationArray, vars, eqns, index, ht)) algorithm
+      DAE.ALGORITHM_STMTS(statementLst=stmts) := alg_;
+      size := -index;
+      allPreStmts := {};
+      allStmts := {};
+      for stmt in stmts loop
+        (stmts, preStmts, index) := encapsulateWhenConditions_Algorithms({stmt}, vars, index);
```
Comment (sjoelund) [inline Compiler/BackEnd/FindZeroCrossings.mo]: Couldn't you just listReverse(stmts) or fix encapsulateWhenConditions_Algorithms instead?

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2017-03-09_15-43-03](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/3999/console).

**Comment by vwaurich** [comment]
encapsulateWhenConditions_Algorithms() outputs preStmts if there is a when equation inside an algorithm. This preStmt contains the assignment encapsulation of when conditions. If there are asserts in the stmts, encapsulateWhenConditions_Algorithms() no preStmt is created although the whenCondition variable is added. Hence, the system is under determined. The for-loop fixes the issue without changing the function interface.

**Comment by vwaurich** [comment]
@lochel OpenHydraulics.DevelopmentTests.LossyPumpTestSimple

File: Compiler/BackEnd/FindZeroCrossings.mo (hunk 1 above)
Comment (vwaurich) [inline Compiler/BackEnd/FindZeroCrossings.mo]: If coverage is fine, I will check if I can simplify encapsulateWhenConditions_Algorithms furthermore

**Comment by lochel** [comment]
@vwaurich I see! Thanks for fixing this.

**Comment by lochel** [comment]
Coverage improved for [OpenHydraulics](https://trac.openmodelica.org/OpenModelica/wiki/CoverageIssues#OpenHydraulics) 👍

**Comment by vwaurich** [comment]
@lochel Yeah, this was a cool bug ;) Did you notice any coverage decrease?

**Comment by lochel** [comment]
No, I didn't see any decrease. Everything looks good.

**Pull request closed without merging**